### PR TITLE
Merge multiple subscriptions

### PR DESCRIPTION
**Pull request by mansimarkaur**
A lot of changes need to be made as a lot of tests are failing.

### CODE REVIEW

**Comment by Natim** [comment]
Based on #26 we should merge it and rebase this branch before reviewing.

File: webpush_channels/tests/test_views_subscription.py [hunk 1]
```diff
@@ -79,21 +80,24 @@ def test_create_a_subscription_update_collection_timestamp(self):
...
 
-    def test_create_a_subscription_without_id_generates_a_uuid(self):
+    def test_create_a_subscription_without_id_generates_sha_on_endpoint(self):
         resp = self.app.post_json(self.collection_url,
                                   MINIMALIST_SUBSCRIPTION,
                                   headers=self.headers,
                                   status=201)
         regexp = re.compile(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-'
                             r'[0-9a-f]{4}-[0-9a-f]{12}$')
+        sha = hashlib.sha224(json.dumps(resp.json['data']['endpoint'])).hexdigest()
```
Comment (Natim) [inline webpush_channels/tests/test_views_subscription.py]: Please use SHA-256

File: webpush_channels/tests/test_views_subscription.py [hunk 2]
```diff
@@ -108,7 +112,7 @@ def test_create_a_subscription_with_an_existing_id_returns_existing(self):
         resp = self.app.post_json(self.collection_url,
                                   MINIMALIST_SUBSCRIPTION,
                                   headers=self.headers,
-                                  status=201)
+                                  status=400)
```
Comment (Natim) [inline webpush_channels/tests/test_views_subscription.py]: Here we should be checking for a 200.

File: webpush_channels/tests/test_views_channel.py [hunk 3]
```diff
@@ -114,7 +115,7 @@ def setUp(self):
         self.webpusher_error_patcher = mock.patch('webpush_channels.views.channels.WebPusher')
 
     def test_push_notifications_can_be_sent_to_channel_with_registration_and_subscription(self):
-        data = "You have a message"
+        data = None
```
Comment (Natim) [inline webpush_channels/tests/test_views_channel.py]: Be aware that `json.dumps(None)` will return 'null'. Is it the message we want to send through as the WebPush payload?

File: webpush_channels/views/subscriptions.py [hunk 4]
```diff
@@ -13,8 +16,30 @@ class SubscriptionSchema(ResourceSchema):
...
+        self.config = config
+        self._regexp = None
+        self.request = request.request
+
+        if not self.match(self()):
+            error_msg = "Generated record id does comply with regexp."
+            raise ValueError(error_msg)
+
+    def __call__(self):
+        return hashlib.sha224(json.dumps(self.request.json['data']['endpoint'])).hexdigest()
```
Comment (Natim) [inline webpush_channels/views/subscriptions.py]: Please use sha256 here.

File: webpush_channels/views/subscriptions.py [hunk 5]
```diff
@@ -13,8 +16,30 @@ class SubscriptionSchema(ResourceSchema):
...
 @register(name='subscription',
           collection_path='/subscriptions',
           record_path='/subscriptions/{{id}}')
 class Subscription(UserResource):
     schema = SubscriptionSchema
+
+    def __init__(self, request, context=None):
+        super(Subscription, self).__init__(request, context)
+        if 'id' not in self.request.json['data']:
+            self.model.id_generator = SHAendpoint(self, self.request)
```
Comment (Natim) [inline webpush_channels/views/subscriptions.py]: You don't need to do that in a if, you should be able to just define model.id_generator = SHAendpoint

**Review by Natim — COMMENTED** [review]
That's a good start.

You should add test to define how it should work for:

- Went you post a subscription without an ID
- Went you put a subscription with a wrong ID

Maybe we should not allow a PUT on the subscription endpoint?

File: setup.py [hunk 6]
```diff
@@ -16,8 +16,8 @@ def read_file(filename):
 REQUIREMENTS = [
     'waitress',
     'kinto[postgresql]',
```
Comment (Natim) [inline setup.py]: You should also add `kinto[postgresql] < 6`

File: webpush_channels/views/subscriptions.py [hunk 7]
```diff
@@ -14,15 +21,53 @@ class KeySchema(colander.MappingSchema):
...
+        record = super(SubscriptionSchema, self).deserialize(cstruct)
+        given_id = record.get('id')
+        generated_id = generate_id(record['endpoint'])
+
+        if given_id and given_id != generated_id:
+            raise colander.Invalid(
+                self, msg="Invalid ID: '{}' found while it should be '{}'".format(
+                    given_id, generated_id))
+
+        return super(SubscriptionSchema, self).deserialize(cstruct)
```
Comment (gabisurita) [inline webpush_channels/views/subscriptions.py:42]: Nit: Isn't it the same as returning record here?

Comment (Natim) [reply webpush_channels/views/subscriptions.py:42]: Yes we should change that

**Pull request merged**